File: onlyskills-repo/topological_genus.pl
% Topological Genus Classification - Prolog

:- module(topological_genus, [
    genus/2,
    euler_characteristic/4,
    betti_numbers/4,
    classify_genus/2,
    good_genus/1
]).

% Good genus values
good_genus(0).   % Sphere - topologically simple
good_genus(71).  % Monster - special structure

% Genus classification
genus_class(0, sphere).
genus_class(1, torus).
genus_class(71, monster).
genus_class(G, low) :- G >= 2, G =< 10.
genus_class(G, medium) :- G >= 11, G =< 30.
genus_class(G, high) :- G >= 31, G =< 70.

% Euler characteristic: χ = V - E + F
euler_characteristic(Vertices, Edges, Faces, Chi) :-
    Chi is Vertices - Edges + Faces.

% Genus from Euler characteristic: g = (2 - χ) / 2
genus_from_euler(Chi, Genus) :-
    Genus is abs(2 - Chi) // 2.

% Betti numbers for a surface of genus g
% b0 = connected components (always 1 for a file)
% b1 = 1-dimensional holes (= 2g for closed surface)
% b2 = 2-dimensional voids (= 1 for closed surface)
betti_numbers(Genus, B0, B1, B2) :-
    B0 = 1,
    B1 is 2 * Genus,
    (Genus > 0 -> B2 = 1 ; B2 = 0).

% Classify file by genus
classify_genus(Genus, Classification) :-
    (   Genus = 0 ->
        Classification = good(sphere, zone(11))
    ;   Genus = 71 ->
        Classification = good(monster, zone(71))
    ;   Genus = 1 ->
        Classification = ok(torus, zone(23))
    ;   Genus >= 2, Genus =< 10 ->
        Classification = medium(low_genus, zone(31))
    ;   Genus >= 11, Genus =< 30 ->
        Classification = high(medium_genus, zone(47))
    ;   Genus >= 31, Genus =< 70 ->
        Classification = critical(high_genus, zone(59))
    ;   Classification = unknown(zone(47))
    ).

% Topological invariants
topological_invariant(genus, Genus) :-
    good_genus(Genus).

topological_invariant(euler, Chi) :-
    genus_from_euler(Chi, Genus),
    good_genus(Genus).

topological_invariant(betti, (B0, B1, B2)) :-
    betti_numbers(Genus, B0, B1, B2),
    good_genus(Genus).

% File topology analysis
analyze_file_topology(Path, Inode, Size, Analysis) :-
    % Calculate vertices from path structure
    atom_chars(Path, Chars),
    include(=('/'), Chars, Slashes),
    length(Slashes, SlashCount),
    include(=('.'), Chars, Dots),
    length(Dots, DotCount),
    Vertices is SlashCount + DotCount + 1,
    
    % Calculate edges from inode and size
    Edges is (Inode mod 100) + (Size mod 100),
    
    % Calculate faces from size
    Faces is Size mod 71,
    
    % Compute Euler characteristic
    euler_characteristic(Vertices, Edges, Faces, Chi),
    
    % Compute genus
    genus_from_euler(Chi, Genus),
    
    % Compute Betti numbers
    betti_numbers(Genus, B0, B1, B2),
    
    % Classify
    classify_genus(Genus, Classification),
    
    Analysis = topology(
        genus(Genus),
        euler(Chi),
        betti(B0, B1, B2),
        classification(Classification)
    ).

% Query good files
good_files(Files) :-
    findall(File, (
        file_topology(File, Genus, _, _, _),
        good_genus(Genus)
    ), Files).

% Query by genus range
files_by_genus_range(Min, Max, Files) :-
    findall(File, (
        file_topology(File, Genus, _, _, _),
        Genus >= Min,
        Genus =< Max
    ), Files).

% Topological equivalence
topologically_equivalent(File1, File2) :-
    file_topology(File1, Genus, _, _, _),
    file_topology(File2, Genus, _, _, _).

% Example file records (would be loaded from topological_genus.parquet)
file_topology('/home/user/sphere.rs', 0, 2, (1, 0, 0), 11).
file_topology('/home/user/monster.lean', 71, -138, (1, 142, 1), 71).
file_topology('/home/user/torus.pl', 1, 0, (1, 2, 1), 23).

% Example queries:
% ?- good_genus(G).
% ?- classify_genus(0, C).
% ?- classify_genus(71, C).
% ?- analyze_file_topology('/home/test.rs', 12345, 1024, Analysis).
% ?- good_files(Files).
% ?- topologically_equivalent(F1, F2).

File: onlyskills-repo/src/bin/topological_genus.rs
// Topological Genus Classification for Files
use std::path::Path;
use std::fs;
use walkdir::WalkDir;
use polars::prelude::*;
use std::collections::HashMap;

#[derive(Debug, Clone)]
struct FileGenus {
    path: String,
    inode: u64,
    size: u64,
    genus: u8,
    euler_characteristic: i32,
    betti_numbers: (u32, u32, u32),  // b0, b1, b2
    zk71_zone: u8,
}

fn calculate_genus(inode: u64, size: u64, path: &str) -> u8 {
    // Genus from topological invariants
    // genus = (2 - euler_characteristic) / 2
    
    // Calculate Euler characteristic from file properties
    let euler = calculate_euler_characteristic(inode, size, path);
    
    // genus = (2 - χ) / 2
    let genus = ((2 - euler).abs() / 2) as u8;
    
    // Normalize to 0-71 range
    genus % 72
}

fn calculate_euler_characteristic(inode: u64, size: u64, path: &str) -> i32 {
    // χ = V - E + F (vertices - edges + faces)
    // For files, we compute from structure
    
    let vertices = count_vertices(path);
    let edges = count_edges(inode, size);
    let faces = count_faces(size);
    
    (vertices as i32) - (edges as i32) + (faces as i32)
}

fn count_vertices(path: &str) -> u32 {
    // Vertices = path components + extension points
    let components = path.split('/').count() as u32;
    let dots = path.matches('.').count() as u32;
    components + dots
}

fn count_edges(inode: u64, size: u64) -> u32 {
    // Edges = connections between components
    // Based on inode and size relationships
    ((inode % 100) + (size % 100)) as u32
}

fn count_faces(size: u64) -> u32 {
    // Faces = enclosed regions
    // Derived from file size structure
    (size % 71) as u32
}

fn calculate_betti_numbers(inode: u64, size: u64, genus: u8) -> (u32, u32, u32) {
    // Betti numbers: b0 (components), b1 (holes), b2 (voids)
    
    // b0 = number of connected components (always 1 for a file)
    let b0 = 1;
    
    // b1 = number of 1-dimensional holes (genus)
    let b1 = genus as u32;
    
    // b2 = number of 2-dimensional voids
    let b2 = if genus > 0 { 1 } else { 0 };
    
    (b0, b1, b2)
}

fn classify_by_genus(genus: u8) -> u8 {
    // ZK71 zone based on genus
    match genus {
        0 => 11,      // Genus 0 = sphere = GOOD (Zone 11)
        71 => 71,     // Genus 71 = Monster = GOOD (Zone 71)
        1 => 23,      // Genus 1 = torus = OK (Zone 23)
        2..=10 => 31, // Low genus = MEDIUM (Zone 31)
        11..=30 => 47, // Medium genus = HIGH (Zone 47)
        31..=70 => 59, // High genus = CRITICAL (Zone 59)
        _ => 47,      // Unknown = HIGH (Zone 47)
    }
}

fn scan_files_with_genus(base: &Path, limit: usize) -> Vec<FileGenus> {
    WalkDir::new(base)
        .max_depth(5)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().is_file())
        .take(limit)
        .filter_map(|entry| {
            let path = entry.path();
            let path_str = path.display().to_string();
            
            let metadata = entry.metadata().ok()?;
            let inode = std::os::unix::fs::MetadataExt::ino(&metadata);
            let size = metadata.len();
            
            let genus = calculate_genus(inode, size, &path_str);
            let euler = calculate_euler_characteristic(inode, size, &path_str);
            let betti = calculate_betti_numbers(inode, size, genus);
            let zone = classify_by_genus(genus);
            
            Some(FileGenus {
                path: path_str,
                inode,
                size,
                genus,
                euler_characteristic: euler,
                betti_numbers: betti,
                zk71_zone: zone,
            })
        })
        .collect()
}

fn main() {
    println!("🔄 Topological Genus Classification");
    println!("{}", "=".repeat(70));
    println!();
    
    let base = Path::new("/home/mdupont/experiments/monster");
    let limit = 10_000;
    
    println!("📊 Scanning {} files...", limit);
    let files = scan_files_with_genus(base, limit);
    println!("✓ Scanned {} files", files.len());
    println!();
    
    // Create DataFrame
    let df = DataFrame::new(vec![
        Series::new("path", files.iter().map(|f| f.path.clone()).collect::<Vec<_>>()),
        Series::new("inode", files.iter().map(|f| f.inode).collect::<Vec<_>>()),
        Series::new("size", files.iter().map(|f| f.size).collect::<Vec<_>>()),
        Series::new("genus", files.iter().map(|f| f.genus).collect::<Vec<_>>()),
        Series::new("euler_characteristic", files.iter().map(|f| f.euler_characteristic).collect::<Vec<_>>()),
        Series::new("betti_0", files.iter().map(|f| f.betti_numbers.0).collect::<Vec<_>>()),
        Series::new("betti_1", files.iter().map(|f| f.betti_numbers.1).collect::<Vec<_>>()),
        Series::new("betti_2", files.iter().map(|f| f.betti_numbers.2).collect::<Vec<_>>()),
        Series::new("zk71_zone", files.iter().map(|f| f.zk71_zone).collect::<Vec<_>>()),
    ]).unwrap();
    
    println!("📋 Sample data:");
    println!("{}", df.head(Some(5)));
    println!();
    
    // Write to Parquet
    let mut file = fs::File::create("topological_genus.parquet").unwrap();
    ParquetWriter::new(&mut file).finish(&mut df.clone()).unwrap();
    
    println!("✓ Saved: topological_genus.parquet");
    println!();
    
    // Statistics by genus
    println!("📊 Distribution by Genus:");
    let genus_0 = files.iter().filter(|f| f.genus == 0).count();
    let genus_71 = files.iter().filter(|f| f.genus == 71).count();
    let genus_1 = files.iter().filter(|f| f.genus == 1).count();
    
    println!("  Genus 0 (sphere): {} files - GOOD (Zone 11)", genus_0);
    println!("  Genus 71 (Monster): {} files - GOOD (Zone 71)", genus_71);
    println!("  Genus 1 (torus): {} files - OK (Zone 23)", genus_1);
    
    let mut genus_counts: HashMap<u8, usize> = HashMap::new();
    for file in &files {
        *genus_counts.entry(file.genus).or_insert(0) += 1;
    }
    
    let mut sorted: Vec<_> = genus_counts.into_iter().collect();
    sorted.sort_by(|a, b| b.1.cmp(&a.1));
    
    println!();
    println!("📊 Top 10 Genus Values:");
    for (genus, count) in sorted.iter().take(10) {
        let pct = *count as f64 / files.len() as f64 * 100.0;
        println!("  Genus {}: {} files ({:.1}%)", genus, count, pct);
    }
    
    println!();
    println!("📊 Distribution by ZK71 Zone:");
    for zone in [11, 23, 31, 47, 59, 71] {
        let count = files.iter().filter(|f| f.zk71_zone == zone).count();
        let pct = count as f64 / files.len() as f64 * 100.0;
        
        let label = match zone {
            11 => "Genus 0 (sphere)",
            71 => "Genus 71 (Monster)",
            23 => "Genus 1 (torus)",
            31 => "Low genus",
            47 => "Medium genus",
            59 => "High genus",
            _ => "Unknown",
        };
        
        println!("  Zone {}: {} files ({:.1}%) - {}", zone, count, pct, label);
    }
    
    println!();
    println!("∞ Genus 0 = Good. Genus 71 = Good. Topology Invariants. ∞");
}
